### PR TITLE
[feature](nereids) Integrate nereids into current SQL process framework

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Integrate nereids into current SQL process framework.

## Problem Summary:

To integrate the nereids optimizer with new SQLParser and Planner to existing SQL process framework, I abstract a interface which named "Planner" and let the Both planner from nereids and stale optimizer implement it, to distinguish  with origin Planner, I rename the Planner to OriginalPlanner.

As we don't want to impact the existing logic too much, I defined a LogicalPlanAdapter to adapt the logicalPlan that is the output of the new paser to the existing code.

Besides, as the MySQL  protocol supports sending multiple statements in one packet, so I add `Nereids#SparseSQL` method to handle this properly.

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
4. Has unit tests been added: (No)
5. Has document been added or modified: (No Need)
6. Does it need to update dependencies: (No)
7. Are there any changes that cannot be rolled back: (No)

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/NereidsParser.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        return statementBaseList;
+    }
+
+    /**
+     * parse sql DSL string.
+     *
+     * @param sql sql string
+     * @return logical plan
+     */
+    public LogicalPlan parseSingle(String sql) throws Exception {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/NereidsParser.java:64]: moving in front of `parseMultiple` is better

**Review by morrySnow — DISMISSED** [review]
LGTM

**Review by morrySnow — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlannerAdapter.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.doris.qe.ConnectContext;
+import org.apache.doris.thrift.TQueryOptions;
+
+import java.util.ArrayList;
+
+/**
+ * This class is used for the compatibility and code reuse in.
+ * @see org.apache.doris.qe.StmtExecutor
+ */
+public class PlannerAdapter extends Planner {
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlannerAdapter.java:38]: Whether to abstract the original planner out of an interface for compatibility between the old and new optimizers. Let both the old and new optimizer planner inherit from this interface. One is NereidsPlanner and the other is OriginPlanner.
Each implements its own plan function, and the return value is List<PlanFragment> .

At present, although this adaptor mode can be implemented, the abstraction is not good.

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/PlannerAdapter.java:38]: Okay, I'll refactor it in the future

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlannerAdapter.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            throw new RuntimeException("Wrong type of queryStmt, expected: <? extends LogicalPlanAdapter>");
+        }
+        LogicalPlanAdapter logicalPlanAdapter = (LogicalPlanAdapter) queryStmt;
+        PhysicalPlan physicalPlan = planner.plan(logicalPlanAdapter.getLogicalPlan(), new PhysicalProperties(), ctx);
+        PhysicalPlanTranslator physicalPlanTranslator = new PhysicalPlanTranslator();
+        PlanContext planContext = new PlanContext();
+        physicalPlanTranslator.translatePlan(physicalPlan, planContext);
+        fragments = new ArrayList<>(planContext.getPlanFragmentList());
+        PlanFragment root = fragments.get(fragments.size() - 1);
+        root.getPlanRoot().convertToVectoriezd();
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlannerAdapter.java:66]: You may need to add ```if (VectorizedUtil.isVectorized()) ``` here

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 4]
```diff
@@ -735,8 +737,12 @@ private void analyzeAndGenerateQueryPlan(TQueryOptions tQueryOptions) throws Use
         }
         plannerProfile.setQueryAnalysisFinishTime();
 
-        // create plan
-        planner = new Planner();
+        if (parsedStmt instanceof LogicalPlanAdapter) {
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:740]:
```
if (parsedStmt instanceof LogicalPlanAdapter) {
    planner = NereidsPlanner();
} else {
    planner = OriginPlanner();
}
```

File: fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java [hunk 5]
```diff
@@ -6,430 +6,47 @@
...
-        }
-        return singleNodePlanner.getScanNodes();
-    }
-
-    public void plan(StatementBase queryStmt, Analyzer analyzer, TQueryOptions queryOptions)
-            throws UserException {
-        createPlanFragments(queryStmt, analyzer, queryOptions);
-    }
+    public abstract void plan(StatementBase queryStmt,
+            org.apache.doris.thrift.TQueryOptions queryOptions) throws UserException;
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java:38]:
```suggestion
            TQueryOptions queryOptions) throws UserException;
```

File: fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java [hunk 6]
```diff
@@ -6,430 +6,47 @@
...
-                str.append("\n");
-            }
-            str.append("PLAN FRAGMENT " + i + "\n");
-            str.append(fragment.getExplainString(explainLevel));
-        }
-        if (explainLevel == TExplainLevel.VERBOSE) {
-            str.append(plannerContext.getRootAnalyzer().getDescTbl().getExplainString());
-        }
-        return str.toString();
+        return "Not implemented yet";
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java [hunk 7]
```diff
@@ -6,430 +6,47 @@
...
-                    }
-                }
-            }
-        }
-    }
-
-    /**
-     * Return combined explain string for all plan fragments.
-     */
     public String getExplainString(List<PlanFragment> fragments, ExplainOptions explainOptions) {
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java (hunk 5 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/Planner.java]: done

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**